Split output filetype variable into atm and sfc

WalterKolczynski-NOAA opened the pull request:
The determination of the FV3 output filetype was a bit unwieldy as it
contained two different settings (one for the atmosphere and one for
the surface) that were not necessarily changed at the same time. They
were also being determined in a different location than other settings
based on the model resolution.

Now the old OUTPUT_FILETYPES variable has been split into two different
variables, OUTPUT_FILETYPES_ATM and OUTPUT_FILETYPES_SFC. The determi-
nation was also moved into the config.fv3 file, where other resolution-
dependent computational settings are set.

This has resulted in some functional change on WCOSS Dell, as the EnKF
was using different chunking settings and never used parallel output
for the surface likely because wasn't being run at a low enough resolu-
tion to consider it. However, there is no reason to believe the two
forecast modes should have different chunk or output settings. So, I've
used the EnKF chunking settings and the free forecast switchover point
for the surface output mode. However, on Hera the EnKF still forces 
serial output, as there is some issue with parallel netcdf and ensembles 
on that machine.

The new filetype settings are now also divorced from the OUTPUT_FILE
setting. However, that setting will soon be unnecessary as nemsio
is removed from the code as an option (see Issue #601).

Fixes: #600
Refs: #601

**Type of change**
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)

**How Has This Been Tested?**
- [x] Cycled test on Hera C384/C196
- [x] Cylced test on Orion C384/C196
- [x] Cycled test on WCOSS Dell C768/C384
  
**Checklist**
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] My changes generate no new warnings
- [x] New and existing tests pass with my changes
- [x] Any dependent changes have been merged and published

WalterKolczynski-NOAA committed:
Split output filetype variable into atm and sfc

The determination of the FV3 output filetype was a bit unwieldy as it
contained two different settings (one for the atmosphere and one for
the surface) that were not necessarily changed at the same time. They
were also being determined in a different location than other settings
based on the model resolution.

Now the old OUTPUT_FILETYPES variable has been split into two different
variables, OUTPUT_FILETYPES_ATM and OUTPUT_FILETYPES_SFC. The determi-
nation was also moved into the config.fv3 file, where other resolution-
dependent computational settings are set.

This has resulted in some functional change on WCOSS Dell, as the EnKF
was using different chunking settings and never used parallel output
for the surface likely because wasn't being run at a low enough resolu-
tion to consider it. However, there is no reason to believe the two
forecast modes should have different chunk or output settings. So, I've
used the EnKF chunking settings and the free forecast switchover point
for the surface output mode.

The new filetype settings are now also divorced from the OUTPUT_FILE
setting. However, that setting will soon be unnecessary as nemsio
is removed from the code as an option (see Issue #601).

Fixes: #600
Refs: #601

WalterKolczynski-NOAA commented:
Leaving as draft until I repeat the test on WCOSS Dell.

Adding GDAS team as reviewers since this technically effects them.

WalterKolczynski-NOAA commented:
Actually, I think I need to run this again at a higher resolution since these are so resolution dependent. I only ran the same C384/C196 test I've run previously. I don't anticipate any code changes though.

junwang-noaa commented on ush/parsing_model_configure_FV3.sh:
```diff
@@ -38,7 +38,7 @@ write_nsflip:            ${WRITE_NSFLIP:-".false."}
 filename_base:           'atm' 'sfc'
 output_grid:             $OUTPUT_GRID
-output_file:             $OUTPUT_FILETYPES
+output_file:             '$OUTPUT_FILETYPE_ATM' '$OUTPUT_FILETYPE_SFC'
```
These two settings are resolution dependent, "parallel_netcdf" is not recommended to lower resolution than C384L127 as it takes longer. For C384L127, the "parallel_netcdf" runs faster for atmf files, but not for sfcf file for C768L127, it is recommended to used "parallel_netcdf" for both atmf and sfcf files. If the workflow will also be used by low resolution runs, especially long forecast run, then using "netcdf" will be a better choice.

aerorahul replied:
@junwang-noaa 
We are choosing to defining these variables based on the resolution in `config.fv3`. 
There is a `default` also set at the top there.

junwang-noaa commented on parm/config/config.fv3:
```diff
@@ -101,6 +105,17 @@ case $case_in in
+        fi
+        if [[ $machine == "WCOSS_DELL_P3" ]]; then
+            RESTILE=$(echo $CASE |cut -c 2-)
+            export ichunk2d=$((4*RESTILE))
```
May I ask what is the reason to choose these specific chunksizes? What benefit do we get?

aerorahul replied:
They are just being moved from `config.fcst` to `config.fv3`

WalterKolczynski-NOAA replied:
All chunk settings were pulled from what was already being used in the EnKF or free forecast previously.

junwang-noaa commented:
I remember I had a discussion with Kate. The ichunk2d should be same as the
output domain IMO/JMO, which is equal to the (4*RESTILE) and (2*RESTILE)
at this time, but it would be good to set inchunk2d/ichunk3d as "IMO"
directly., same for JMO

On Thu, Jan 20, 2022 at 8:38 AM Rahul Mahajan ***@***.***>
wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In parm/config/config.fv3
> <https://github.com/NOAA-EMC/global-workflow/pull/602#discussion_r788771502>
> :
>
> > @@ -101,6 +105,17 @@ case $case_in in
>          export WRITE_GROUP_GFS=${WRITE_GROUP_GFS:-2}
>          export WRTTASK_PER_GROUP_GFS=${WRTTASK_PER_GROUP_GFS:-$npe_node_max}
>          export WRTIOBUF=${WRTIOBUF:-"16M"}
> +        if [[ $machine == "ORION" ]]; then
> +            export OUTPUT_FILETYPE_SFC="netcdf_parallel"
> +        fi
> +        if [[ $machine == "WCOSS_DELL_P3" ]]; then
> +            RESTILE=$(echo $CASE |cut -c 2-)
> +            export ichunk2d=$((4*RESTILE))
>
> They are just being moved from config.fcst to config.fv3
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/NOAA-EMC/global-workflow/pull/602#discussion_r788771502>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AI7D6TORKXOCTYUTGDK3A6TUXAF6ZANCNFSM5MMC66VQ>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
> You are receiving this because you commented.Message ID:
> ***@***.***>
>

aerorahul committed:
separate into 2 case blocks

WalterKolczynski-NOAA commented:
> I remember I had a discussion with Kate. The ichunk2d should be same as the output domain IMO/JMO, which is equal to the (4*RESTILE) and (2*RESTILE) at this time, but it would be good to set inchunk2d/ichunk3d as "IMO" directly., same for JMO

If this is true for all resolutions, we can just pull all that out of the CASE block and set it in one place.

WalterKolczynski-NOAA committed:
Set chunksize for all resolutions

Based on conversations with Jun and others, the chunksize should be
set using the same formula for all resolutions and machines.

WalterKolczynski-NOAA committed:
Force serial I/O for EnKF on Hera

There is an issue with parallel netcdf specifically for ensemble runs
on Hera, so code to force serial output there is restored to the EnKF
config.

Refs: #600

WalterKolczynski-NOAA commented:
System now runs on all three core machines.

aerorahul reviewed (approved):
Looks good. 
Approve pending testing.

WalterKolczynski-NOAA commented:
> Looks good. Approve pending testing.

Testing is already complete. Still hoping someone from DA will sign off since there is a difference in chunking and serial/parallel at some resolutions.

WalterKolczynski-NOAA commented:
@RussTreadon-NOAA @CatherineThomas-NOAA can one of you give this a quick look before I merge?

CoryMartin-NOAA commented:
Were any timing tests performed with new vs old chunk sizes for both forecast and GSI? I seem to remember a great deal of variability in both write-grid component write speed and GSI read speed based on the choices selected for chunk size.

See: https://docs.google.com/spreadsheets/d/1DPXJ6-3OXzn67-7ZK_VRttDl1kwYPhf-FBndtg9ZAnM/edit?usp=sharing from @RussTreadon-NOAA

WalterKolczynski-NOAA commented:
> Were any timing tests performed with new vs old chunk sizes for both forecast and GSI? I seem to remember a great deal of variability in both write-grid component write speed and GSI read speed based on the choices selected for chunk size.
> 
> See: https://docs.google.com/spreadsheets/d/1DPXJ6-3OXzn67-7ZK_VRttDl1kwYPhf-FBndtg9ZAnM/edit?usp=sharing from @RussTreadon-NOAA

I didn't do any timing tests, but the 3d chunking should be unchanged at operational resolutions. This is mostly extending the same chunking rubric to coarser resolutions and using the 2d chunk rubric EnKF was already doing in other forecast jobs.

CatherineThomas-NOAA reviewed (commented):
The changes look okay to me, though I'm not knowledgeable about chunking and will defer to someone who knows more from the DA side (@CoryMartin-NOAA? @jswhit2?)
